File: llamafile/server/www/highlight_shell.js
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

const SHELL_KEYWORDS = new Set([
  'break',
  'case',
  'coproc',
  'do',
  'done',
  'elif',
  'else',
  'esac',
  'exec',
  'exit',
  'expr',
  'fi',
  'for',
  'function',
  'if',
  'in',
  'return',
  'select',
  'then',
  'time',
  'trap',
  'until',
  'while',
]);

const SHELL_BUILTINS = new Set([
  'alias',
  'bg',
  'bind',
  'builtin',
  'caller',
  'cd',
  'chdir',
  'command',
  'declare',
  'echo',
  'enable',
  'eval',
  'false',
  'fg',
  'getopts',
  'hash',
  'help',
  'jobs',
  'kill',
  'let',
  'local',
  'logout',
  'mapfile',
  'printf',
  'read',
  'readarray',
  'set',
  'shift',
  'source',
  'stop',
  'suspend',
  'test',
  'times',
  'true',
  'type',
  'typeset',
  'ulimit',
  'unalias',
  'unset',
  'wait',
]);

class HighlightShell extends Highlighter {

  static NORMAL = 0;
  static WORD = 1;
  static QUOTE = 2;
  static DQUOTE = 3;
  static DQUOTE_VAR = 4;
  static DQUOTE_VAR2 = 5;
  static DQUOTE_CURL = 6;
  static DQUOTE_CURL_BACKSLASH = 7;
  static DQUOTE_BACKSLASH = 8;
  static TICK = 9;
  static TICK_BACKSLASH = 10;
  static VAR = 11;
  static VAR2 = 12;
  static CURL = 13;
  static CURL_BACKSLASH = 14;
  static COMMENT = 15;
  static LT = 16;
  static LT_LT = 17;
  static LT_LT_NAME = 18;
  static LT_LT_QNAME = 19;
  static HEREDOC_BOL = 20;
  static HEREDOC = 21;
  static HEREDOC_VAR = 22;
  static HEREDOC_VAR2 = 23;
  static HEREDOC_CURL = 24;
  static HEREDOC_CURL_BACKSLASH = 25;
  static BACKSLASH = 26;

  constructor(delegate) {
    super(delegate);
    this.c = 0;
    this.i = 0;
    this.curl = 0;
    this.last = 0;
    this.pending_heredoc = false;
    this.indented_heredoc = false;
    this.no_interpolation = false;
    this.heredoc = '';
    this.word = '';
  }

  feed(input) {
    for (let i = 0; i < input.length; i += this.delta) {
      this.delta = 1;
      let c = input[i];
      this.last = this.c;
      this.c = c;
      switch (this.state) {

      case HighlightShell.NORMAL:
        if (!isascii(c) || isalpha(c) || c == '_') {
          this.state = HighlightShell.WORD;
          this.word += c;
        } else if (c == '\'') {
          this.state = HighlightShell.QUOTE;
          this.push("span", "string");
          this.append('\'');
        } else if (c == '\\') {
          this.state = HighlightShell.BACKSLASH;
          this.push("span", "escape");
          this.append('\\');
        } else if (c == '"') {
          this.state = HighlightShell.DQUOTE;
          this.push("span", "string");
          this.append('"');
        } else if (c == '`') {
          this.state = HighlightShell.TICK;
          this.push("span", "string");
          this.append('`');
        } else if (c == '$') {
          this.state = HighlightShell.VAR;
          this.append('$');
        } else if (c == '<') {
          this.state = HighlightShell.LT;
          this.append('<');
        } else if (c == '#' && (!this.last || isspace(this.last))) {
          this.push("span", "comment");
          this.append('#');
          this.state = HighlightShell.COMMENT;
        } else if (c == '\n') {
          this.append('\n');
          if (this.pending_heredoc) {
            this.push("span", "string");
            this.pending_heredoc = false;
            this.state = HighlightShell.HEREDOC_BOL;
            this.i = 0;
          }
        } else {
          this.append(c);
        }
        break;

      case HighlightShell.BACKSLASH:
        this.append(c);
        this.pop();
        this.state = HighlightShell.NORMAL;
        break;

      case HighlightShell.WORD:
        if (!isascii(c) || isalnum(c) || c == '_') {
          this.word += c;
        } else {
          if (SHELL_KEYWORDS.has(this.word)) {
            this.push("span", "keyword");
            this.append(this.word);
            this.pop();
          } else if (SHELL_BUILTINS.has(this.word)) {
            this.push("span", "builtin");
            this.append(this.word);
            this.pop();
          } else {
            this.append(this.word);
          }
          this.word = '';
          this.epsilon(HighlightShell.NORMAL);
        }
        break;

      case HighlightShell.VAR:
        if (c == '!' || //
            c == '#' || //
            c == '$' || //
            c == '*' || //
            c == '-' || //
            c == '?' || //
            c == '@' || //
            c == '\\' || //
            c == '^') {
          this.push("span", "var");
          this.append(c);
          this.pop();
          this.state = HighlightShell.NORMAL;
          break;
        } else if (c == '{') {
          this.append('{');
          this.push("span", "var");
          this.state = HighlightShell.CURL;
          this.curl = 1;
          break;
        } else {
          this.push("span", "var");
          this.state = HighlightShell.VAR2;
        }
        // fallthrough

      case HighlightShell.VAR2:
        if (!isascii(c) || isalnum(c) || c == '_') {
          this.append(c);
        } else {
          this.pop();
          this.epsilon(HighlightShell.NORMAL);
        }
        break;

      case HighlightShell.CURL:
        if (c == '\\') {
          this.state = HighlightShell.CURL_BACKSLASH;
          this.pop();
          this.push("span", "escape");
          this.append('\\');
        } else if (c == '{') {
          this.pop();
          this.append('{');
          this.push("span", "var");
          ++this.curl;
        } else if (c == '}') {
          this.pop();
          this.append('}');
          if (!--this.curl) {
            this.state = HighlightShell.NORMAL;
          }
        } else if (ispunct(c)) {
          this.pop();
          this.append(c);
        } else {
          this.append(c);
        }
        break;

      case HighlightShell.CURL_BACKSLASH:
        this.append(c);
        this.pop();
        this.state = HighlightShell.CURL;
        break;

      case HighlightShell.COMMENT:
        this.append(c);
        if (c == '\n') {
          this.pop();
          this.state = HighlightShell.NORMAL;
        }
        break;

      case HighlightShell.QUOTE:
        this.append(c);
        if (c == '\'') {
          this.pop();
          this.state = HighlightShell.NORMAL;
        }
        break;

      case HighlightShell.DQUOTE:
        if (c == '"') {
          this.append(c);
          this.pop();
          this.state = HighlightShell.NORMAL;
        } else if (c == '\\') {
          this.append(c);
          this.state = HighlightShell.DQUOTE_BACKSLASH;
        } else if (c == '$') {
          this.state = HighlightShell.DQUOTE_VAR;
        } else {
          this.append(c);
        }
        break;

      case HighlightShell.DQUOTE_BACKSLASH:
        this.append(c);
        this.state = HighlightShell.DQUOTE;
        break;

      case HighlightShell.DQUOTE_VAR:
        if (c == '!' || //
            c == '#' || //
            c == '$' || //
            c == '*' || //
            c == '-' || //
            c == '?' || //
            c == '@' || //
            c == '\\' || //
            c == '^') {
          this.push("span", "bold");
          this.append('$');
          this.append(c);
          this.pop();
          this.state = HighlightShell.DQUOTE;
          break;
        } else if (c == '{') {
          this.push("span", "bold");
          this.append("${");
          this.state = HighlightShell.DQUOTE_CURL;
          this.curl = 1;
          break;
        } else if (c == '(') {
          this.append('$');
          this.state = HighlightShell.DQUOTE_VAR2;
        } else {
          this.push("span", "bold");
          this.append('$');
          this.state = HighlightShell.DQUOTE_VAR2;
        }
        // fallthrough

      case HighlightShell.DQUOTE_VAR2:
        if (!isascii(c) || isalnum(c) || c == '_') {
          this.append(c);
        } else {
          this.pop();
          this.epsilon(HighlightShell.DQUOTE);
        }
        break;

      case HighlightShell.DQUOTE_CURL:
        if (c == '\\') {
          this.state = HighlightShell.DQUOTE_CURL_BACKSLASH;
          this.append('\\');
        } else if (c == '{') {
          this.append('{');
          ++this.curl;
        } else if (c == '}') {
          this.append('}');
          if (!--this.curl) {
            this.pop();
            this.state = HighlightShell.DQUOTE;
          }
        } else if (ispunct(c)) {
          this.append(c);
        } else {
          this.append(c);
        }
        break;

      case HighlightShell.DQUOTE_CURL_BACKSLASH:
        this.append(c);
        this.state = HighlightShell.DQUOTE_CURL;
        break;

      case HighlightShell.TICK:
        this.append(c);
        if (c == '`') {
          this.pop();
          this.state = HighlightShell.NORMAL;
        } else if (c == '\\') {
          this.state = HighlightShell.TICK_BACKSLASH;
        }
        break;

      case HighlightShell.TICK_BACKSLASH:
        this.append(c);
        this.state = HighlightShell.TICK;
        break;

      case HighlightShell.LT:
        if (c == '<') {
          this.append(c);
          this.state = HighlightShell.LT_LT;
          this.heredoc = '';
          this.pending_heredoc = false;
          this.indented_heredoc = false;
          this.no_interpolation = false;
        } else {
          this.epsilon(HighlightShell.NORMAL);
        }
        break;

      case HighlightShell.LT_LT:
        if (c == '-') {
          this.indented_heredoc = true;
          this.append(c);
        } else if (c == '\\') {
          this.append(c);
        } else if (c == '\'') {
          this.state = HighlightShell.LT_LT_QNAME;
          this.push("span", "string");
          this.append(c);
          this.no_interpolation = true;
        } else if (isalpha(c) || c == '_') {
          this.state = HighlightShell.LT_LT_NAME;
          this.heredoc += c;
          this.append(c);
        } else if (isascii(c) && isblank(c)) {
          this.append(c);
        } else {
          this.epsilon(HighlightShell.NORMAL);
        }
        break;

      case HighlightShell.LT_LT_NAME:
        if (isalnum(c) || c == '_') {
          this.state = HighlightShell.LT_LT_NAME;
          this.heredoc += c;
          this.append(c);
        } else if (c == '\n') {
          this.append(c);
          this.push("span", "string");
          this.state = HighlightShell.HEREDOC_BOL;
        } else {
          this.pending_heredoc = true;
          this.epsilon(HighlightShell.NORMAL);
        }
        break;

      case HighlightShell.LT_LT_QNAME:
        this.append(c);
        if (c == '\'') {
          this.pop();
          this.state = HighlightShell.HEREDOC_BOL;
          this.pending_heredoc = true;
          this.state = HighlightShell.NORMAL;
        } else {
          this.heredoc += c;
        }
        break;

      case HighlightShell.HEREDOC_BOL:
        this.append(c);
        if (c == '\n') {
          if (this.i == this.heredoc.length) {
            this.state = HighlightShell.NORMAL;
            this.pop();
          }
          this.i = 0;
        } else if (c == '\t' && this.indented_heredoc) {
          // do nothing
        } else if (this.i < this.heredoc.length && this.heredoc[this.i] == c) {
          this.i++;
        } else {
          this.state = HighlightShell.HEREDOC;
          this.i = 0;
        }
        break;

      case HighlightShell.HEREDOC:
        if (c == '\n') {
          this.append('\n');
          this.state = HighlightShell.HEREDOC_BOL;
        } else if (c == '$' && !this.no_interpolation) {
          this.state = HighlightShell.HEREDOC_VAR;
        } else {
          this.append(c);
        }
        break;

      case HighlightShell.HEREDOC_VAR:
        if (c == '!' || //
            c == '#' || //
            c == '$' || //
            c == '*' || //
            c == '-' || //
            c == '?' || //
            c == '@' || //
            c == '\\' || //
            c == '^') {
          this.push("span", "bold");
          this.append('$');
          this.append(c);
          this.pop();
          this.state = HighlightShell.HEREDOC;
          break;
        } else if (c == '{') {
          this.push("span", "bold");
          this.append("${");
          this.state = HighlightShell.HEREDOC_CURL;
          this.curl = 1;
          break;
        } else if (c == '(') {
          this.append('$');
          this.state = HighlightShell.HEREDOC_VAR2;
        } else {
          this.push("span", "bold");
          this.append('$');
          this.state = HighlightShell.HEREDOC_VAR2;
        }
        // fallthrough

      case HighlightShell.HEREDOC_VAR2:
        if (!isascii(c) || isalnum(c) || c == '_') {
          this.append(c);
        } else {
          this.pop();
          this.epsilon(HighlightShell.HEREDOC);
        }
        break;

      case HighlightShell.HEREDOC_CURL:
        if (c == '\\') {
          this.state = HighlightShell.HEREDOC_CURL_BACKSLASH;
          this.append('\\');
        } else if (c == '{') {
          this.append('{');
          ++this.curl;
        } else if (c == '}') {
          this.append('}');
          if (!--this.curl) {
            this.pop();
            this.state = HighlightShell.HEREDOC;
          }
        } else if (ispunct(c)) {
          this.append(c);
        } else {
          this.append(c);
        }
        break;

      case HighlightShell.HEREDOC_CURL_BACKSLASH:
        this.append(c);
        this.state = HighlightShell.HEREDOC_CURL;
        break;

      default:
        throw new Error('Invalid state');
      }
    }
  }

  flush() {
    switch (this.state) {
    case HighlightShell.WORD:
      if (SHELL_KEYWORDS.has(this.word)) {
        this.push("span", "keyword");
        this.append(this.word);
        this.pop();
      } else if (SHELL_BUILTINS.has(this.word)) {
        this.push("span", "builtin");
        this.append(this.word);
        this.pop();
      } else {
        this.append(this.word);
      }
      this.word = '';
      break;
    case HighlightShell.DQUOTE_VAR:
      this.append('$');
      this.pop();
      break;
    case HighlightShell.HEREDOC_VAR:
      this.append('$');
      this.pop();
      break;
    case HighlightShell.VAR2:
    case HighlightShell.CURL:
    case HighlightShell.CURL_BACKSLASH:
    case HighlightShell.TICK:
    case HighlightShell.TICK_BACKSLASH:
    case HighlightShell.QUOTE:
    case HighlightShell.DQUOTE:
    case HighlightShell.DQUOTE_VAR2:
    case HighlightShell.DQUOTE_CURL:
    case HighlightShell.DQUOTE_CURL_BACKSLASH:
    case HighlightShell.DQUOTE_BACKSLASH:
    case HighlightShell.COMMENT:
    case HighlightShell.HEREDOC_BOL:
    case HighlightShell.HEREDOC:
    case HighlightShell.HEREDOC_VAR2:
    case HighlightShell.HEREDOC_CURL:
    case HighlightShell.HEREDOC_CURL_BACKSLASH:
    case HighlightShell.LT_LT_QNAME:
    case HighlightShell.BACKSLASH:
      this.pop();
      break;
    default:
      break;
    }
    this.state = HighlightShell.NORMAL;
    this.last = 0;
    this.delegate.flush();
    this.delta = 1;
  }
}

Highlighter.REGISTRY['shell'] = HighlightShell;
Highlighter.REGISTRY['bash'] = HighlightShell;
Highlighter.REGISTRY['ksh'] = HighlightShell;
Highlighter.REGISTRY['sh'] = HighlightShell;

File: llamafile/highlight/highlight_shell.cpp
// -*- mode:c++;indent-tabs-mode:nil;c-basic-offset:4;coding:utf-8 -*-
// vi: set et ft=cpp ts=4 sts=4 sw=4 fenc=utf-8 :vi
//
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "highlight.h"
#include "util.h"
#include <cosmo.h>
#include <ctype.h>

enum
{
    NORMAL,
    WORD,
    QUOTE,
    DQUOTE,
    DQUOTE_VAR,
    DQUOTE_VAR2,
    DQUOTE_CURL,
    DQUOTE_CURL_BACKSLASH,
    DQUOTE_BACKSLASH,
    TICK,
    TICK_BACKSLASH,
    VAR,
    VAR2,
    CURL,
    CURL_BACKSLASH,
    COMMENT,
    LT,
    LT_LT,
    LT_LT_NAME,
    LT_LT_QNAME,
    HEREDOC_BOL,
    HEREDOC,
    HEREDOC_VAR,
    HEREDOC_VAR2,
    HEREDOC_CURL,
    HEREDOC_CURL_BACKSLASH,
    BACKSLASH,
};

HighlightShell::HighlightShell()
{
}

HighlightShell::~HighlightShell()
{
}

void
HighlightShell::feed(std::string* r, std::string_view input)
{
    for (size_t i = 0; i < input.size(); ++i) {
        wchar_t c;
        int b = input[i] & 255;
        last_ = c_;
        if (!u_) {
            if (b < 0300) {
                c_ = c = b;
            } else {
                c_ = ThomPikeByte(b);
                u_ = ThomPikeLen(b) - 1;
                continue;
            }
        } else if (ThomPikeCont(b)) {
            c = c_ = ThomPikeMerge(c_, b);
            if (--u_)
                continue;
        } else {
            u_ = 0;
            c = b;
        }
        if (c == 0xFEFF)
            continue; // utf-8 bom
        switch (t_) {

        Normal:
        case NORMAL:
            if (!isascii(c) || isalpha(c) || c == '_') {
                t_ = WORD;
                lf::append_wchar(&word_, c);
            } else if (c == '\'') {
                t_ = QUOTE;
                *r += HI_STRING;
                *r += '\'';
            } else if (c == '\\') {
                t_ = BACKSLASH;
                *r += HI_ESCAPE;
                *r += '\\';
            } else if (c == '"') {
                t_ = DQUOTE;
                *r += HI_STRING;
                *r += '"';
            } else if (c == '`') {
                t_ = TICK;
                *r += HI_STRING;
                *r += '`';
            } else if (c == '$') {
                t_ = VAR;
                *r += '$';
            } else if (c == '<') {
                t_ = LT;
                *r += '<';
            } else if (c == '#' && (!last_ || isspace(last_))) {
                *r += HI_COMMENT;
                *r += '#';
                t_ = COMMENT;
            } else if (c == '\n') {
                *r += '\n';
                if (pending_heredoc_) {
                    *r += HI_STRING;
                    pending_heredoc_ = false;
                    t_ = HEREDOC_BOL;
                    i_ = 0;
                }
            } else {
                lf::append_wchar(r, c);
            }
            break;

        case BACKSLASH:
            lf::append_wchar(r, c);
            *r += HI_RESET;
            t_ = NORMAL;
            break;

        case WORD:
            if (!isascii(c) || isalnum(c) || c == '_') {
                lf::append_wchar(&word_, c);
            } else {
                if (is_keyword_shell(word_.data(), word_.size())) {
                    *r += HI_KEYWORD;
                    *r += word_;
                    *r += HI_RESET;
                } else if (is_keyword_shell_builtin(word_.data(),
                                                    word_.size())) {
                    *r += HI_BUILTIN;
                    *r += word_;
                    *r += HI_RESET;
                } else {
                    *r += word_;
                }
                word_.clear();
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case VAR:
            if (c == '!' || //
                c == '#' || //
                c == '$' || //
                c == '*' || //
                c == '-' || //
                c == '?' || //
                c == '@' || //
                c == '\\' || //
                c == '^') {
                *r += HI_VAR;
                lf::append_wchar(r, c);
                *r += HI_RESET;
                t_ = NORMAL;
                break;
            } else if (c == '{') {
                *r += '{';
                *r += HI_VAR;
                t_ = CURL;
                curl_ = 1;
                break;
            } else {
                *r += HI_VAR;
                t_ = VAR2;
            }
            // fallthrough

        case VAR2:
            if (!isascii(c) || isalnum(c) || c == '_') {
                lf::append_wchar(r, c);
            } else {
                *r += HI_RESET;
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case CURL:
            if (c == '\\') {
                t_ = CURL_BACKSLASH;
                *r += HI_RESET;
                *r += HI_ESCAPE;
                *r += '\\';
            } else if (c == '{') {
                *r += HI_RESET;
                *r += '{';
                *r += HI_VAR;
                ++curl_;
            } else if (c == '}') {
                *r += HI_RESET;
                *r += '}';
                if (!--curl_) {
                    t_ = NORMAL;
                }
            } else if (ispunct(c)) {
                *r += HI_RESET;
                lf::append_wchar(r, c);
            } else {
                lf::append_wchar(r, c);
            }
            break;

        case CURL_BACKSLASH:
            lf::append_wchar(r, c);
            *r += HI_RESET;
            t_ = CURL;
            break;

        case COMMENT:
            lf::append_wchar(r, c);
            if (c == '\n') {
                *r += HI_RESET;
                t_ = NORMAL;
            }
            break;

        case QUOTE:
            lf::append_wchar(r, c);
            if (c == '\'') {
                *r += HI_RESET;
                t_ = NORMAL;
            }
            break;

        Dquote:
        case DQUOTE:
            if (c == '"') {
                lf::append_wchar(r, c);
                *r += HI_RESET;
                t_ = NORMAL;
            } else if (c == '\\') {
                lf::append_wchar(r, c);
                t_ = DQUOTE_BACKSLASH;
            } else if (c == '$') {
                t_ = DQUOTE_VAR;
            } else {
                lf::append_wchar(r, c);
            }
            break;

        case DQUOTE_BACKSLASH:
            lf::append_wchar(r, c);
            t_ = DQUOTE;
            break;

        case DQUOTE_VAR:
            if (c == '!' || //
                c == '#' || //
                c == '$' || //
                c == '*' || //
                c == '-' || //
                c == '?' || //
                c == '@' || //
                c == '\\' || //
                c == '^') {
                *r += HI_BOLD;
                *r += '$';
                lf::append_wchar(r, c);
                *r += HI_UNBOLD;
                t_ = DQUOTE;
                break;
            } else if (c == '{') {
                *r += HI_BOLD;
                *r += "${";
                t_ = DQUOTE_CURL;
                curl_ = 1;
                break;
            } else if (c == '(') {
                *r += '$';
                t_ = DQUOTE_VAR2;
            } else {
                *r += HI_BOLD;
                *r += '$';
                t_ = DQUOTE_VAR2;
            }
            // fallthrough

        case DQUOTE_VAR2:
            if (!isascii(c) || isalnum(c) || c == '_') {
                lf::append_wchar(r, c);
            } else {
                *r += HI_UNBOLD;
                t_ = DQUOTE;
                goto Dquote;
            }
            break;

        case DQUOTE_CURL:
            if (c == '\\') {
                t_ = DQUOTE_CURL_BACKSLASH;
                *r += '\\';
            } else if (c == '{') {
                *r += '{';
                ++curl_;
            } else if (c == '}') {
                *r += '}';
                if (!--curl_) {
                    *r += HI_UNBOLD;
                    t_ = DQUOTE;
                }
            } else if (ispunct(c)) {
                lf::append_wchar(r, c);
            } else {
                lf::append_wchar(r, c);
            }
            break;

        case DQUOTE_CURL_BACKSLASH:
            lf::append_wchar(r, c);
            t_ = DQUOTE_CURL;
            break;

        case TICK:
            lf::append_wchar(r, c);
            if (c == '`') {
                *r += HI_RESET;
                t_ = NORMAL;
            } else if (c == '\\') {
                t_ = TICK_BACKSLASH;
            }
            break;

        case TICK_BACKSLASH:
            lf::append_wchar(r, c);
            t_ = TICK;
            break;

        case LT:
            if (c == '<') {
                lf::append_wchar(r, c);
                t_ = LT_LT;
                heredoc_.clear();
                pending_heredoc_ = false;
                indented_heredoc_ = false;
                no_interpolation_ = false;
            } else {
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case LT_LT:
            if (c == '-') {
                indented_heredoc_ = true;
                lf::append_wchar(r, c);
            } else if (c == '\\') {
                lf::append_wchar(r, c);
            } else if (c == '\'') {
                t_ = LT_LT_QNAME;
                *r += HI_STRING;
                lf::append_wchar(r, c);
                no_interpolation_ = true;
            } else if (isalpha(c) || c == '_') {
                t_ = LT_LT_NAME;
                lf::append_wchar(&heredoc_, c);
                lf::append_wchar(r, c);
            } else if (isascii(c) && isblank(c)) {
                *r += c;
            } else {
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case LT_LT_NAME:
            if (isalnum(c) || c == '_') {
                t_ = LT_LT_NAME;
                lf::append_wchar(&heredoc_, c);
                lf::append_wchar(r, c);
            } else if (c == '\n') {
                lf::append_wchar(r, c);
                *r += HI_STRING;
                t_ = HEREDOC_BOL;
            } else {
                pending_heredoc_ = true;
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case LT_LT_QNAME:
            lf::append_wchar(r, c);
            if (c == '\'') {
                *r += HI_RESET;
                t_ = HEREDOC_BOL;
                pending_heredoc_ = true;
                t_ = NORMAL;
            } else {
                lf::append_wchar(&heredoc_, c);
            }
            break;

        case HEREDOC_BOL:
            lf::append_wchar(r, c);
            if (c == '\n') {
                if (i_ == heredoc_.size()) {
                    t_ = NORMAL;
                    *r += HI_RESET;
                }
                i_ = 0;
            } else if (c == '\t' && indented_heredoc_) {
                // do nothing
            } else if (i_ < heredoc_.size() && (heredoc_[i_] & 255) == c) {
                i_++;
            } else {
                t_ = HEREDOC;
                i_ = 0;
            }
            break;

        Heredoc:
        case HEREDOC:
            if (c == '\n') {
                *r += '\n';
                t_ = HEREDOC_BOL;
            } else if (c == '$' && !no_interpolation_) {
                t_ = HEREDOC_VAR;
            } else {
                lf::append_wchar(r, c);
            }
            break;

        case HEREDOC_VAR:
            if (c == '!' || //
                c == '#' || //
                c == '$' || //
                c == '*' || //
                c == '-' || //
                c == '?' || //
                c == '@' || //
                c == '\\' || //
                c == '^') {
                *r += HI_BOLD;
                *r += '$';
                lf::append_wchar(r, c);
                *r += HI_UNBOLD;
                t_ = HEREDOC;
                break;
            } else if (c == '{') {
                *r += HI_BOLD;
                *r += "${";
                t_ = HEREDOC_CURL;
                curl_ = 1;
                break;
            } else if (c == '(') {
                *r += '$';
                t_ = HEREDOC_VAR2;
            } else {
                *r += HI_BOLD;
                *r += '$';
                t_ = HEREDOC_VAR2;
            }
            // fallthrough

        case HEREDOC_VAR2:
            if (!isascii(c) || isalnum(c) || c == '_') {
                lf::append_wchar(r, c);
            } else {
                *r += HI_UNBOLD;
                t_ = HEREDOC;
                goto Heredoc;
            }
            break;

        case HEREDOC_CURL:
            if (c == '\\') {
                t_ = HEREDOC_CURL_BACKSLASH;
                *r += '\\';
            } else if (c == '{') {
                *r += '{';
                ++curl_;
            } else if (c == '}') {
                *r += '}';
                if (!--curl_) {
                    *r += HI_UNBOLD;
                    t_ = HEREDOC;
                }
            } else if (ispunct(c)) {
                lf::append_wchar(r, c);
            } else {
                lf::append_wchar(r, c);
            }
            break;

        case HEREDOC_CURL_BACKSLASH:
            lf::append_wchar(r, c);
            t_ = HEREDOC_CURL;
            break;

        default:
            __builtin_unreachable();
        }
    }
}

void
HighlightShell::flush(std::string* r)
{
    switch (t_) {
    case WORD:
        if (is_keyword_shell(word_.data(), word_.size())) {
            *r += HI_KEYWORD;
            *r += word_;
            *r += HI_RESET;
        } else if (is_keyword_shell_builtin(word_.data(), word_.size())) {
            *r += HI_BUILTIN;
            *r += word_;
            *r += HI_RESET;
        } else {
            *r += word_;
        }
        word_.clear();
        break;
    case DQUOTE_VAR:
        *r += '$';
        *r += HI_RESET;
        break;
    case HEREDOC_VAR:
        *r += '$';
        *r += HI_RESET;
        break;
    case VAR2:
    case CURL:
    case CURL_BACKSLASH:
    case TICK:
    case TICK_BACKSLASH:
    case QUOTE:
    case DQUOTE:
    case DQUOTE_VAR2:
    case DQUOTE_CURL:
    case DQUOTE_CURL_BACKSLASH:
    case DQUOTE_BACKSLASH:
    case COMMENT:
    case HEREDOC_BOL:
    case HEREDOC:
    case HEREDOC_VAR2:
    case HEREDOC_CURL:
    case HEREDOC_CURL_BACKSLASH:
    case LT_LT_QNAME:
    case BACKSLASH:
        *r += HI_RESET;
        break;
    default:
        break;
    }
    c_ = 0;
    u_ = 0;
    t_ = NORMAL;
    last_ = 0;
}
